android/src/main/java/io/ably/flutter/plugin/push/PushActivationEventHandlers.java

```diff
@@ -1,106 +1,97 @@
 package io.ably.flutter.plugin.push;
 
-import android.content.Context;
-import android.content.Intent;
-import android.content.IntentFilter;
+import static io.ably.flutter.plugin.push.PushActivationReceiver.PUSH_ACTIVATE_ACTION;
+import static io.ably.flutter.plugin.push.PushActivationReceiver.PUSH_DEACTIVATE_ACTION;
+import static io.ably.flutter.plugin.push.PushActivationReceiver.PUSH_UPDATE_FAILED_ACTION;
 
+import android.content.Context;
 import android.util.Log;
+
 import androidx.annotation.Nullable;
-import androidx.localbroadcastmanager.content.LocalBroadcastManager;
 
 import io.ably.flutter.plugin.generated.PlatformConstants;
 import io.ably.lib.types.ErrorInfo;
-import io.ably.lib.util.IntentUtils;
 import io.flutter.plugin.common.MethodChannel;
 
 public class PushActivationEventHandlers {
-  private static final String TAG = PushActivationEventHandlers.class.getName();
-  private static final String PUSH_ACTIVATE_ACTION = "io.ably.broadcast.PUSH_ACTIVATE";
-  private static final String PUSH_DEACTIVATE_ACTION = "io.ably.broadcast.PUSH_DEACTIVATE";
-  private static final String PUSH_UPDATE_FAILED_ACTION = "io.ably.broadcast.PUSH_UPDATE_FAILED";
-
-  static PushActivationEventHandlers instance;
-  final private BroadcastReceiver broadcastReceiver;
-  final private MethodChannel methodChannel;
-  private MethodChannel.Result resultForActivate;
-  private MethodChannel.Result resultForDeactivate;
-
-  public static void instantiate(Context context, MethodChannel methodChannel) {
-    if (instance == null) {
-      instance = new PushActivationEventHandlers(context, methodChannel);
+    private static final String TAG = PushActivationEventHandlers.class.getName();
+
+    private static PushActivationEventHandlers instance;
+
+    private MethodChannel.Result resultForActivate;
+    private MethodChannel.Result resultForDeactivate;
+
+    private final PushActivationReceiver pushActivationReceiver;
+    private final MethodChannel methodChannel;
+    private final Context context;
+
+    private PushActivationEventHandlers(Context context, MethodChannel methodChannel) {
+        this.methodChannel = methodChannel;
+        this.context = context;
+        pushActivationReceiver = new PushActivationReceiver();
     }
-  }
 
-  public static void setResultForActivate(MethodChannel.Result result) {
-    instance.resultForActivate = result;
-  }
+    public static synchronized void instantiate(Context context, MethodChannel methodChannel) {
+        if (instance == null) {
+            instance = new PushActivationEventHandlers(context, methodChannel);
+        }
+    }
 
-  public static void setResultForDeactivate(MethodChannel.Result result) {
-    instance.resultForDeactivate = result;
-  }
+    public static PushActivationEventHandlers getInstance() {
+        return instance;
+    }
```

Comment:
This can lead to NPEs since you need to first call `instantiate()` before calling this method to assign some value to the `instance` field. However, I see that this is how it has worked before so I wouldn't want to block the PR on this thing :wink:

Reply:
You are right and you are right 😄 This should also be protected by the its use case. Meaning only when flutter engine is attached (meaning methods will be able to communicate over a channel) this will be called. But this will most probably be history for next couple of changes for push notifications